File: MediumDifficulty/SortAnArray.cpp
class Solution {
public:
    void merge(vector<int>& arr, int l, int m, int r)
    {
        vector<int> left(m - l + 1);
        for(auto i {l}; i <= m; ++i)
            left[i - l] = arr[i];
        
        vector<int> right(r - m);
        for(auto j {m + 1}; j <= r; ++j)
            right[j  - (m + 1)] = arr[j];
        
        int i = 0;
        int j = 0;
        int k = l;
        while(i < left.size() && j < right.size())
        {
            if(left[i] <= right[j])
            {
                arr[k] = left[i];
                ++i;
            }
            else if(right[j] < left[i])
            {
                arr[k] = right[j];
                ++j;
            }
            ++k;
        }
        
        while(i < left.size())
        {
            arr[k] = left[i];
            ++i;
            ++k;
        }
        
        while(j < right.size())
        {
            arr[k] = right[j];
            ++j;
            ++k;
        }
    }
    
    void mergeSort(vector<int>& nums, int l, int r)
    {
        if(l < r)
        {
            int m = (l + r) / 2;
            mergeSort(nums, l, m);
            mergeSort(nums, m + 1, r);
            merge(nums, l, m, r);
        }
    }
    
    vector<int> sortArray(vector<int>& nums) {
        int l = 0;
        int r = nums.size() - 1;
        mergeSort(nums, l, r);
        
        return nums;
    }
};

File: Python/medium_difficulty/SortAnArray.py
class Solution:
    def merge(self, arr: List[int], l: int, m: int, r: int):
        left = arr[l : m + 1]
        right = arr[m + 1 : r + 1]

        i = 0
        j = 0
        k = l
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                arr[k] = left[i]
                i += 1
            elif right[j] < left[i]:
                arr[k] = right[j]
                j += 1
            k += 1

        while i < len(left):
            arr[k] = left[i]
            i += 1
            k += 1

        while j < len(right):
            arr[k] = right[j]
            j += 1
            k += 1

    def mergeSort(self, nums: List[int], l: int, r: int):
        if l < r:
            m = (l + r) // 2

            self.mergeSort(nums, l, m)
            self.mergeSort(nums, m + 1, r)
            self.merge(nums, l, m, r)

    def sortArray(self, nums: List[int]) -> List[int]:
        l = 0
        r = len(nums) - 1

        self.mergeSort(nums, l, r)

        return nums